packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -738,6 +746,14 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             }
         }
 
+        const attributorProvider: FluidObject<IProvideAttributorProvider> = containerScope;
+        const attributor = context.audience ? await attributorProvider.IAttributorProvider?.initialize(
```

Comment:
i wouldn't expect any changes in the container runtime itself. i would expect all changes to be done optionally in a container runtime factory, which passes scope to this method

Reply:
look at how aqueduct injects dependency synthesizer: 
https://github.com/microsoft/FluidFramework/blob/abc1703691c505f5932059a00b08bc87dece4d6e/packages/framework/aqueduct/src/container-runtime-factories/baseContainerRuntimeFactory.ts#L75

Reply:
then in the datastore factory, it looks for the synthesizer on scope: https://github.com/microsoft/FluidFramework/blob/abc1703691c505f5932059a00b08bc87dece4d6e/packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts#L94

to instantiate the services the datastore needs

Reply:
most consumer don't use our aqueduct's container runtime factory, as they need to customize things. they do use the data store factory. But they use the same pattern to inject the dependency synthesize, but will a bunch more serives, attributor could be one such service. 

its not clear to be how we get the blob written. that would probably need to be done via a mixin to the actually container runtime instance knows how to write that specific blob. 

the other options is to wrap attributor in datastore that the container runtime factory creates by default, and then makes available on the scope. The attribute datastore would then be responsible for writing the data on summarize. This was the goal of something like shared summary blob, but i don't know if we ever solved all the orchestration issues.

Reply:
great, thanks for the pointers.

> the other options is to wrap attributor in datastore that the container runtime factory creates by default, and then makes available on the scope. The attribute datastore would then be responsible for writing the data on summarize. This was the goal of something like shared summary blob, but i don't know if we ever solved all the orchestration issues.

AgentScheduler does something like this now, right? Don't think it injects things onto the scope though.

Reply:
I pushed some changes that move things to a more proper mixin model which mirrors the mixin strategy we use for `DataStoreRuntime`. Since ContainerRuntime needs some async loading logic, there are still a couple changes necessary, but it's along the lines of what we do elsewhere (pass constructor params through certain factory classes). Now that I think about it, this work is also necessary to enable de-kludging some of what WB has done in their container runtime implementation. I'm going to write some unit tests and PR those changes in separately.